round-based/src/echo_broadcast/mod.rs

```diff
@@ -0,0 +1,458 @@
+//!
+//! // execute the protocol
+//! let keyshare = keygen(mpc, i, n).await?;
+//! # Ok(()) }
+//! ```
+
+use core::marker::PhantomData;
+
+use alloc::collections::btree_map::BTreeMap;
+use digest::Digest;
+
+use crate::{
+    round::{RoundInfo, RoundStore, RoundStoreExt},
+    Mpc, MpcExecution, Outgoing, ProtocolMsg, RoundMsg,
+};
+
+mod error;
+mod store;
+
+pub use self::error::{CompleteRoundError, EchoError, Error};
+
+/// Message of the protocol with echo broadcast round(s)
+pub enum Msg<D: Digest, M> {
+    /// Message from echo broadcast sub-protocol
+    Echo {
+        /// Indicates for which round of main protocol this echo message is transmitted
+        ///
+        /// Note that this field is controlled by potential malicious party. If it sets it
+        /// to the round that doesn't exist, the protocol will likely be aborted with an error
+        /// that we received a message from unregistered round, which may appear as implementation
+        /// error (i.e. API misuse), but in fact it's a malicious abort.
+        round: u16,
+        /// Hash of all messages received in `round`
+        hash: digest::Output<D>,
+    },
+    /// Message from the main protocol
+    Main(M),
+}
+
+/// Sub-messages of [`Msg`]
+///
+/// Sub-messages implement [`RoundMsg`] trait for [`Msg`]
+mod sub_msg {
+    pub struct EchoMsg<D: digest::Digest, R> {
+        pub hash: digest::Output<D>,
+        pub _round: core::marker::PhantomData<R>,
+    }
+    #[derive(Debug, Clone)]
+    pub struct Main<M>(pub M);
+
+    impl<D: digest::Digest, R> Clone for EchoMsg<D, R> {
+        fn clone(&self) -> Self {
+            Self {
+                hash: self.hash.clone(),
+                _round: core::marker::PhantomData,
+            }
+        }
+    }
+}
+
+impl<D: Digest, M: Clone> Clone for Msg<D, M> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Echo { round, hash } => Self::Echo {
+                round: *round,
+                hash: hash.clone(),
+            },
+            Self::Main(msg) => Self::Main(msg.clone()),
+        }
+    }
+}
+
+impl<D: Digest, M: PartialEq> PartialEq for Msg<D, M> {
+    fn eq(&self, other: &Self) -> bool {
+        match self {
+            Self::Echo { round, hash } => {
+                matches!(other, Self::Echo { round: r2, hash: h2 } if round == r2 && hash == h2)
+            }
+            Self::Main(msg) => matches!(other, Self::Main(m2) if msg == m2),
+        }
+    }
+}
+
+impl<D: Digest, M: PartialEq> Eq for Msg<D, M> {}
+
+impl<D: Digest, M: core::fmt::Debug> core::fmt::Debug for Msg<D, M> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Self::Echo { round, hash } => f
+                .debug_struct("Msg::Echo")
+                .field("round", round)
+                .field("hash", hash)
+                .finish(),
+            Self::Main(msg) => f.debug_tuple("Msg::Main").field(msg).finish(),
+        }
+    }
+}
```

Comment:
This is obvious after reading, but it would save me several minutes if you added a comment saying "implemented this instance by hand becuase of unfulfilled bound on `D` instead of `digest::Output<D>`"